gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -924,6 +932,218 @@ public void deleteObjects(List<StorageResourceId> fullObjectNames) throws IOExce
+        FolderInfo folderToDelete = folderDeleteBlockingQueue.poll(1L, TimeUnit.MINUTES);
+
+        // if after 1 minute, if the queue is still empty it will return null and loop is exited
+        if (folderToDelete == null) {
+          break;
+        }
+        queueSingleFolderDelete(
+            folderToDelete,
+            countOfChildren,
+            folderDeleteBlockingQueue,
+            innerExceptions,
+            executor,
+            1);
+      }
+      // deleting any remaining resources
+      executor.shutdown();
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+
+    if (!innerExceptions.isEmpty()) {
+      GoogleCloudStorageEventBus.postOnException();
+      throw GoogleCloudStorageExceptions.createCompositeException(innerExceptions);
+    }
+  }
+
+  /* Callable class specifically for deletion of folder resource */
+  private class DeleteFolder implements Callable<Void> {
+    private DeleteFolderRequest deleteFolderRequest;
+
+    @Override
+    public Void call() throws IOException {
+      logger.atFiner().log("Deleting folder resource %s", deleteFolderRequest.getName());
+      lazyGetStorageControlClient().deleteFolder(deleteFolderRequest);
+      return null;
+    }
+
+    DeleteFolder(FolderInfo folder) {
+      this.deleteFolderRequest =
+          DeleteFolderRequest.newBuilder().setName(folder.toString()).build();
+    }
+  }
+
+  /** Helper function to delete a single folder resource */
+  public void queueSingleFolderDelete(
+      @Nonnull final FolderInfo folder,
+      ConcurrentHashMap<String, Long> countOfChildren,
+      BlockingQueue<FolderInfo> folderDeleteBlockingQueue,
+      final KeySetView<IOException, Boolean> innerExceptions,
+      final BatchExecutor batchExecutor,
+      final int attempt)
+      throws IOException {
+    final String bucketName = folder.getBucket();
+    final String folderName = folder.getFolderName();
+    checkArgument(
+        !Strings.isNullOrEmpty(bucketName),
+        String.format("No bucket for folder resource %s", folder.toString()));
+    checkArgument(
+        !Strings.isNullOrEmpty(folderName),
+        String.format("No folder path for folder resource %s", folder.toString()));
+
+    batchExecutor.queue(
+        new DeleteFolder(folder),
+        getDeletionCallback(
+            folder,
+            countOfChildren,
+            folderDeleteBlockingQueue,
+            innerExceptions,
+            batchExecutor,
+            attempt));
+  }
+
+  /**
+   * Helper function to add folderResource to blocking queue
+   *
+   * @param folderResource
+   * @param folderDeleteBlockingQueue blocking queue
+   */
+  private void addFolderResourceInBlockingQueue(
+      FolderInfo folderResource,
+      BlockingQueue<FolderInfo> folderDeleteBlockingQueue,
+      KeySetView<IOException, Boolean> innerExceptions) {
+    try {
+      folderDeleteBlockingQueue.add(folderResource);
+    } catch (IllegalStateException e) {
```

Comment:
when can this happen?

Reply:
This can happen if the queue overflows. Ideally it should never happen but implementation restricts to handle the IllegalStateException error.